compiler/rustc_builtin_macros/src/asm.rs

```diff
@@ -424,6 +428,65 @@ fn expand_preparsed_asm(ecx: &mut ExtCtxt<'_>, sp: Span, args: AsmArgs) -> P<ast
 
         let template_str = &template_str.as_str();
         let template_snippet = ecx.source_map().span_to_snippet(template_sp).ok();
+
+        if let Some(snippet) = &template_snippet {
+            let default_dialect = match ecx.sess.asm_arch {
```

Comment:
This probably wants to be deduplicated with a similar piece of code in `rustc_codegen_llvm`. Probably a method inherent to `Session` in `compiler/rustc_session/src/session.rs` would be best.

Reply:
The similar code in `rustc_codegen_llvm/src/asm.rs` in `codegen_inline_asm` specifically uses inline asm options in addition.  I am not sure if that belongs in Session?  I did move the **default** assembly syntax calculation there, as that feels dependent on the session alone, and not on some other options.  Should I create a version that takes options to calculate it as well?